server/main.go

```diff
@@ -30,19 +30,38 @@ func languagesHandler(w http.ResponseWriter, r *http.Request) {
 
 func runHandler(w http.ResponseWriter, r *http.Request) {
 	// TODO: parse request body
+	decoder := json.NewDecoder(r.Body)
+	var res api.RunRequest
+	err := decoder.Decode(&res)
 
+	if err != nil {
```

Comment:
Ok unfortunately this is a Go "footgun" but relatively easy to correct. Decoders should be treated kind of like a stream we have to continue reading until `err == io.EOF` to verify no errors in decoding. When running this, I was confused why I wouldn't get any response and it was because this was silently throwing an error and not decoding my request properly.

This article was helpful: https://dev.to/taqkarim/you-might-not-be-using-json-decoder-correctly-in-golang-12mb

Something like this?

```go
	var err error

	for {
		err = decoder.Decode(&res)
		if err != nil {
			log.Printf("error parsing request: %v\n", err)
		}
		if err == io.EOF {
			break
		}
	}
```

Also, when we do encounter an error, can we return a 400 error code? Something like this:

```go
    // this can be handled inside or outside the for loop and maybe in a function...
	if err != nil {
		w.WriteHeader(http.StatusBadRequest)
		w.Header().Set("Content-Type", "application/json")
		resp := make(map[string]string)
		resp["error"] = "cannot parse request"
		jsonResp, _ := json.Marshal(resp)
		w.Write(jsonResp)
		return
	}
```